### ISSUE TITLE
0^0 should be defined 1

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#605]
**Reported by thorade on 5 Sep 2011 10:56 UTC**
Currently 0<sup>0</sup> will give an error (at least in Dymola), but in technical applications the only reasonable value for 0<sup>0</sup> is 1, as discussed here: http://www.faqs.org/faqs/sci-math-faq/specialnumbers/0to0/

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/605

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 5 Sep 2011 11:32 UTC**
Same issue was discussed in ticket #101

The given reference is basically one person selectively referencing people he agrees with and ignoring the ones saying something else - hardly the best reference.
For a more complete picture see:

http://en.wikipedia.org/wiki/Exponentation#Zero_to_the_zero_power 

note especially that the 0<sup>a</sup> and a<sup>0</sup> are relevant for Modelica.

**Comment by modelica-trac-importer** [comment]
**Comment by thorade on 5 Sep 2011 13:05 UTC**
Sorry, I did not see ticket #101 as I don't have the privileges to see it.
Is there a recommended way how to treat this case?

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 6 Sep 2011 12:17 UTC**
Although #101 was finally marked invalid, there are Modelica tools interpreting this differently. We should probably come up with a rule which all tools can follow so that we don't have differing interpretation of mathematics from one tool to another.

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 12 Dec 2012 11:05 UTC**
Dymola 2013FD01 does now return 0⁰=1, didn't check with other Dymola versions.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 12 Dec 2012 16:30 UTC**
It looks like most tools use 0<sup>0</sup> = 1 (OM, Python and even Dymola now). So perhaps we should just specify it then rather than leaving it unclear.

**Comment by modelica-trac-importer** [comment]
**Changelog removed by dietmarw on 12 Dec 2012 16:30 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 13 Dec 2012 10:33 UTC**
The wikipedia reference should now be:

http://en.wikipedia.org/wiki/Exponentation#Zero_to_the_power_of_zero

However, the situation is still the same, and the question is not only about 0<sup>0</sup>, but also about a<sup>0</sup> and 0<sup>a</sup>; and those are often evaluated in incompatible ways in tools.

This is important since one normally expects that if f(x)=g(x) then f(0)=g(0).

**Comment by modelica-trac-importer** [comment]
**Comment by thorade on 13 Dec 2012 11:04 UTC**
Sorry if my view on this is too naive, but maybe it helps to split this up into two questions. First question is: Should this be defined by the Modelica Spec? I guess this is easy to answer, because all tools should return equal output for equal input. I do not see any advantage in leaving it undefined.

Second question is: How to define it? As far as I know for all a<>0 we have a<sup>0</sup>=1 and 0<sup>a</sup>=0. So if I did not miss anything the only case that is undefined is 0<sup>0</sup>. As discussed in the Wikipedia article and elsewhere, there is not one single correct definition. I see three possible definitions: return an error/undefined value, return 1 or return 0.  Defininig it to be 1 seems to be the least evil and is the way how it is done in "most programming languages" (Wikipedia).

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 13 Dec 2012 12:13 UTC**
Replying to [comment:7 thorade@…]:
> As far as I know for all a<>0 we have a<sup>0</sup>=1 and 0<sup>a</sup>=0.

Not necessarily, it can be argued that 0<sup>a</sup> for a<0 is 1/(0<sup>-a</sup>) = 1/0 and thus a division by zero error.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 13 Dec 2012 12:30 UTC**
Replying to [comment:8 jmattsson]:
> Not necessarily, it can be argued that 0<sup>a</sup> for a<0 is 1/(0<sup>-a</sup>) = 1/0 and thus a division by zero error.

That case is currently in the specification, just like 0<sup>0</sup> is covered. The problem is of course that undefined means "do whatever you want".

> Consequences of exceptional situations, such as (a==0.0 and b<<sup></sup>=0.0, a<0 and b is not an integer) or overflow are undefined

**Comment by modelica-trac-importer** [comment]
**Comment by petar on 13 Dec 2012 12:56 UTC**
Here's how Mathematica deals with this:

```
In[1]:= 0^0                                      0
Power::indet: Indeterminate expression 0  encountered.
Out[1]= Indeterminate

In[2]:= 0^3
Out[2]= 0
In[3]:= 0^-3                      
-3
Power::infy: Infinite expression 0   encountered.
Out[3]= ComplexInfinity
```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 13 Dec 2012 13:01 UTC**
For me this is completely clear: Modelica is a precise modeling language and is not cheating. If something is not well defined, Modelica treats it as not well defined. 0!<sup>0 is mathematically not well defined and therefore tools should raise an error. If the special case of a parameter setting or a function input argument leads to that case, it is best that the tool raises an error and the modeler has to treat this case specially, and then cases will occur where o!</sup>o = 1 or 0 will not be what the modeler expects, but then the tools tells him to handle this appropriately.

Mathematicas handling is nice.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 13 Dec 2012 15:07 UTC**
I agree with Martin.  If it doesn't have a well defined value, it should be an error.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 14 Dec 2012 12:32 UTC**
Peter: you missed some important cases for Mathematica. 
The previously linked wikipedia-page mentions those and says that Wolfram Alpha is the same; so anyone can test it:

To confirm:
0<sup>0</sup>
(indeterminate)

And then the important one: http://www.wolframalpha.com/input/?i=b%5E0
b<sup>0</sup> 
1

This is only logically consistent **if** 0<sup>0</sup> is **undefined** - so one can ignore it when evaluating b<sup>0</sup> or 0<sup>b</sup>; but if 0<sup>0</sup> would be seen as a strict error that needs to be reported then these manipulations would simply be wrong.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Jun 2016 12:59 UTC**
Housekeeping to make it easier to find the active tickets for the specification.

These tickets are all expired, i.e. old with no action - and not clear that any action is needed. In most cases they are enhancements/discussions without any clear result.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 Nov 2016 09:33 UTC**
Can be solved with #2048.
